File: backend/app/services/openrouter.py
from __future__ import annotations

import base64
import json
import re
from typing import Any

from openai import AsyncOpenAI

from app.config import Settings


def get_openrouter_client(settings: Settings) -> AsyncOpenAI | None:
    if not settings.has_openrouter:
        return None
    return AsyncOpenAI(
        api_key=settings.openrouter_api_key,
        base_url=settings.openrouter_base_url,
    )


async def chat_completion(
    settings: Settings,
    messages: list[dict[str, Any]],
    *,
    model: str | None = None,
    temperature: float = 0.2,
    max_tokens: int = 1200,
) -> str:
    client = get_openrouter_client(settings)
    if client is None:
        raise RuntimeError("OpenRouter API key not configured")

    response = await client.chat.completions.create(
        model=model or settings.openrouter_model,
        messages=messages,
        temperature=temperature,
        max_tokens=max_tokens,
    )
    return (response.choices[0].message.content or "").strip()


async def vision_analyze(
    settings: Settings,
    image_bytes: bytes,
    prompt: str,
    *,
    mime: str = "image/jpeg",
) -> str:
    b64 = base64.b64encode(image_bytes).decode("ascii")
    data_url = f"data:{mime};base64,{b64}"
    messages = [
        {
            "role": "user",
            "content": [
                {"type": "text", "text": prompt},
                {"type": "image_url", "image_url": {"url": data_url}},
            ],
        }
    ]
    return await chat_completion(
        settings,
        messages,
        model=settings.openrouter_vision_model,
        temperature=0.1,
        max_tokens=800,
    )


def extract_json(text: str) -> dict[str, Any]:
    text = text.strip()
    try:
        return json.loads(text)
    except json.JSONDecodeError:
        pass

    fence = re.search(r"```(?:json)?\s*([\s\S]*?)```", text)
    if fence:
        return json.loads(fence.group(1).strip())

    start = text.find("{")
    end = text.rfind("}")
    if start >= 0 and end > start:
        return json.loads(text[start : end + 1])

    raise ValueError(f"Could not parse JSON from model response: {text[:200]}")

File: web/src/lib/server/openrouter.ts
import OpenAI from "openai";
import { AppConfig, hasOpenRouter } from "./config";

export function getOpenRouterClient(cfg: AppConfig) {
  if (!hasOpenRouter(cfg)) return null;
  return new OpenAI({
    apiKey: cfg.openrouterApiKey,
    baseURL: cfg.openrouterBaseUrl,
  });
}

export function extractJson(text: string): Record<string, unknown> {
  const trimmed = text.trim();
  try {
    return JSON.parse(trimmed) as Record<string, unknown>;
  } catch {
    /* continue */
  }
  const fence = trimmed.match(/```(?:json)?\s*([\s\S]*?)```/);
  if (fence) return JSON.parse(fence[1].trim()) as Record<string, unknown>;
  const start = trimmed.indexOf("{");
  const end = trimmed.lastIndexOf("}");
  if (start >= 0 && end > start) {
    return JSON.parse(trimmed.slice(start, end + 1)) as Record<string, unknown>;
  }
  throw new Error(`No JSON in model response: ${trimmed.slice(0, 200)}`);
}

export async function chatCompletion(
  cfg: AppConfig,
  messages: OpenAI.Chat.ChatCompletionMessageParam[],
  opts?: { model?: string; temperature?: number; maxTokens?: number }
) {
  const client = getOpenRouterClient(cfg);
  if (!client) throw new Error("OpenRouter API key not configured");
  const res = await client.chat.completions.create({
    model: opts?.model ?? cfg.openrouterModel,
    messages,
    temperature: opts?.temperature ?? 0.2,
    max_tokens: opts?.maxTokens ?? 1200,
  });
  return (res.choices[0]?.message?.content ?? "").trim();
}

export async function visionAnalyze(
  cfg: AppConfig,
  imageBytes: Buffer,
  prompt: string,
  mime = "image/jpeg"
) {
  const b64 = imageBytes.toString("base64");
  return chatCompletion(
    cfg,
    [
      {
        role: "user",
        content: [
          { type: "text", text: prompt },
          { type: "image_url", image_url: { url: `data:${mime};base64,${b64}` } },
        ],
      },
    ],
    { model: cfg.openrouterVisionModel, temperature: 0.1, maxTokens: 800 }
  );
}
